optuna/samplers/_tpe/sampler.py

```diff
@@ -635,13 +694,44 @@ def _get_observation_pairs(
                 param_value = None
             values[param_name].append(param_value)
 
-    return values, scores
+        if get_constraints:
+            assert constraints is not None
+            constraints.append(cast(Sequence[float], trial.system_attrs.get(_CONSTRAINTS_KEY)))
```

Comment:
If `trial.system_attrs` misses the key of `_CONSTRAINTS_KEY`, the result of `cast(Sequence[float], trial.system_attrs.get(_CONSTRAINTS_KEY))` will be None. So, `constraints` can include `None`.

(1) The type of `constraints` should be `Optional[List[Optional[Sequence[float]]]]`.
(2) The handling of constraints after this function returns should be implemented with the consideration that there may be None inside. (I have not checked the handling after this function yet.)

Reply:
Resolved by https://github.com/optuna/optuna/pull/3506/commits/2f1ff5a57ebe8f09013534cfe311978afc9ab14c.